pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,727 @@
+		func(pl *Pool) bool { _, ok := pl.PodIDs[podID]; return ok })
+	if len(pools) == 0 {
+		return nil
+	}
+	return pools[0]
+}
+
+// allocatePool returns a pool allocated for a pod.
+func (p *podpools) allocatePool(pod cache.Pod) *Pool {
+	if pool := p.allocatedPool(pod); pool != nil {
+		return pool
+	}
+	poolDef := p.getPoolDef(pod)
+	if poolDef == nil {
+		return nil
+	}
+	// Try to find a suitable pool and allocate it for the pod.
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool {
+			return poolDef.Name == pl.Def.Name && (pl.Def.MaxPods > len(pl.PodIDs) || pl.Def.MaxPods == 0)
+		})
+	// Sort pools according to pool type fill order so that the
+	// first pool in the list is the preferred one.
+	switch poolDef.FillOrder {
+	case FillBalanced:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) < len(pools[j].PodIDs)
+		})
+	case FillPacked:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) > len(pools[j].PodIDs)
+		})
+	case FillFirstFree:
+		// FirstFree is already the first of the pools list.
+	}
+	if len(pools) == 0 {
+		log.Warn("cannot find free %q pool for pod %q", poolDef.Name, pod.GetName())
+		return nil
+	}
+	// Found a suitable pool. Allocate it for the pod.
+	podID := pod.GetID()
+	pool := pools[0]
+	pool.PodIDs[podID] = []string{}
+	log.Debug("allocated pool %s[%d] for pod %q", pool.Def.Name, pool.Instance, pod.GetName())
+	return pool
+}
+
+// containersInPool returns the number of containers of a pod in a pool.
+func (p *podpools) containersInPool(pod cache.Pod, pool *Pool) int {
+	if cnts, ok := pool.PodIDs[pod.GetID()]; ok {
+		return len(cnts)
+	}
+	return 0
+}
+
+// dumpPool dumps pool contents in detail.
+func (p *podpools) dumpPool(pool *Pool) string {
+	conts := []string{}
+	pods := []string{}
+	for podID, contIDs := range pool.PodIDs {
+		podName := podID
+		if pod, ok := p.cch.LookupPod(podID); ok {
+			podName = pod.GetName()
+		}
+		pods = append(pods, fmt.Sprintf("%s (mCPU: %d, max=%d)", podName, p.getPodMilliCPU(podID), p.podMaxMilliCPU[podID]))
+		for _, contID := range contIDs {
+			if cont, ok := p.cch.LookupContainer(contID); ok {
+				conts = append(conts, cont.PrettyName())
+			} else {
+				conts = append(conts, podName+":"+contID)
+			}
+		}
+	}
+	s := fmt.Sprintf("Pool{Def.Name: %q, Instance: %d, CPUs: %s, Mems: %s, Def.MaxPods: %d, pods: %v, containers:%v}",
+		pool.Def.Name, pool.Instance, pool.CPUs, pool.Mems, pool.Def.MaxPods, pods, conts)
+	return s
+}
+
+// freePool removes an empty pod from a pool
+func (p *podpools) freePool(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	delete(pool.PodIDs, podID)
+	delete(p.podMaxMilliCPU, podID)
+}
+
+// trackPodCPU keeps track on pod's CPU requests.
+func (p *podpools) trackPodCPU(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	current := p.getPodMilliCPU(podID)
+	if max, ok := p.podMaxMilliCPU[podID]; ok {
+		if max < current {
+			p.podMaxMilliCPU[podID] = current
```

Comment:
Hmm, what is the purpose of this `podMaxMilliCPU` map? Wouldn't `getPodMilliCPU()` be enough? AFAIU the resource allocation (requests or limits) may not change or no new containers may be added after the pod object has been created. I'm probably missing something here 🤔

Reply:
Without an mutating webhook, there is no way to make sure that the resource request for any pool-annotated Pod/continer will be consistent with the configuration of that pool type so that parallel accounting in the scheduler does not go haywire. The next best thing one can do is to detect clearly inconsistent/incorrect configuration/usage and log a diagnostic error message about it.

However, (again without forcing the policy to run with cri-resmgr-webhook or something similar) you cannot detect such misconfiguration when the pod is created. You can tell it once all the containers in the pod have been created. With the amount of stock metadata in a pod or a container, you don't know when that moment in time comes. The easiest thing to do is to record the largest ever concurrent allocation within the pod, then cross-check this with the configuration when the pod is deleted. AFAICT, this is the recording part, and the cross-checking counterpart is `validatePodCPU`.

Reply:
Ah yes, of course, makes perfectly sense. Tend to forget that resource accounting problem...

It might be beneficial for future readers to describe that^ in the docstring/comment of `trackPodCPU()`. WDYT?

Reply:
Yes, I think that would be a good idea.

Reply:
Explained this in new comment in `trackPodCPU()`.